### DISCUSSION TITLE
Get tenant routes to filter from SPA fallback

### DISCUSSION THREAD

**Discussion by xperiandri** [OrchardCMS/OrchardCore#15877] [Q&A]
We are trying to run Blazor from Orchard Core theme.
It launches but `UseBlazor` middleware handles all requests and does not allow to access admin or tenants:

https://github.com/aspnet/Blazor/blob/39d2b90f03b037cf4601a056b37ab55fad9daab6/src/Microsoft.AspNetCore.Blazor.Server/Builder/BlazorApplicationBuilderExtensions.cs#L96

https://github.com/aspnet/Blazor/blob/39d2b90f03b037cf4601a056b37ab55fad9daab6/src/Microsoft.AspNetCore.Blazor.Server/Builder/BlazorApplicationBuilderExtensions.cs#L158

How to get a list of tenant routes, theme route and admin route to filter them from SPA fallback?

**Comment by sebastienros** [comment]
@jtkech Can you take a look?

Maybe we need to add some functional tests that will try to access these from the Demo module.

**Comment by jtkech** [comment]
Okay, i will take a look.

I don't know well `Blazor`, where do you call `UseBlazor()`? If it is through the `Configure()` of a module / theme startup, maybe you just need to override and then set the startup `Order` property to a value > 0 so that the `MapWhenMiddleware` is run later, but need to be tried.

After a quick look at the above links, i saw static file middlewares which use physical files providers. So, i'm afraid that it would need more adaptations because our modules / themes embed their static files in their assemblies and there are served through custom file providers.

I remember we did a recent change allowing to add `Mappings` files to the module static files we serve, because somebody wanted to add e.g `.wasm` files, maybe he was trying to do the same thing.

**Comment by xperiandri** [comment]
My colleague will upload a reproduction [here](https://github.com/msproadshow/Orchard/tree/blazor-conference) (branch blazor-conference) around half of the hour.
We call `UseBlazor()` within the module's startup. And it just handles all the requests.

> I remember we did a recent change allowing to add Mappings files to the module static files we serve, because somebody wanted to add e.g .wasm files, maybe he was trying to do the same thing.

It was me. But we moved to **Blazor** due to **Ooui.Forms** lack of necessary features.
But I'm going to evolve both options.

**Comment by sebastienros** [comment]
Sorry, Blazor, I read Razor Pages ;)

**Comment by xperiandri** [comment]
I've updated the link

**Comment by xperiandri** [comment]
@jtkech, have you tried my reproduction?

**Comment by jtkech** [comment]
@xperiandri sorry not yet, i will do it soon.

**Comment by sebastienros** [comment]
Don';t spend too much time on this, Blazor is super not stable, so I don't want any dependency on it in OC. It's fine to know if it works or not, and if so how to make it work, but we should not try to support it yet. You can already create a side project that would use Blazor to query content from OC, like their Server/Client sample.

**Comment by jtkech** [comment]
Okay, no problem, i just did little tests.

- I first needed to remove any reference on blazor packages to be able to run the setup with your conference recipe.

- Then i got an `SPA default page middleware` exception. Maybe i missed something but your main issue being related to routes, i only focused on this.

 - Indeed, when i remove your path checking then i can't access e.g `/admin`.

So, a quick workaround, which could give you some ideas and which seems to work, is to do something like this in your theme startup.

        // So that it is run last
        public override int Order => 100;

        ....

        public override void Configure(IApplicationBuilder app, ...)
        {
            ...
            app.UseWhen(
                ctx => ctx.Response.StatusCode == StatusCodes.Status404NotFound,
                builder => builder.UseBlazor<RoadShow.Program>());
        }

Doing this i got the same SPA error on the home page (meaning the spa middleware run as before) but i could go e.g to `/admin`.

**Comment by sebastienros** [comment]
I asked around and it seems that mixing Blazor and an MVC app doesn't work, so it would not work with Orchard either. The current samples expose two different apps for this. Let's follow what they have to say on the issue you created in the blazor repos.

**Comment by jtkech** [comment]
@xperiandri have you tried my workaround to filter routes from SPA fallback?

**Comment by xperiandri** [comment]
Yes, it works. Thanks.
Blazor causes an issue now.
